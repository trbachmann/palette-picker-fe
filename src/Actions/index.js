export const setProjects = (projects) => ({
  type: 'SET_PROJECTS',
  projects
});

export const addProjectPalettes = (projectPalettes, id) => ({
  type: 'ADD_PROJECT_PALETTES',
  id,
  projectPalettes
})

export const setError = (error) => ({
  type: 'SET_ERROR',
  error
});

export const toggleLoading = (bool) => ({
  type: 'TOGGLE_LOADING',
  isLoading: bool
});

export const setPaletteColors = (colors) => ({
  type: 'SET_PALETTE_COLORS',
  palette: colors
});

export const setCurrentExpandedProject = (id) => ({
  type: 'SET_CURRENT_EXPANDED_PROJECT',
  id
})
export const addProject = (name, id) => ({
  type: 'ADD_PROJECT',
  name,
  id
});

export const addPaletteToProject = (projectId, palette) => ({
  type: 'ADD_PALETTE',
  projectId,
  palette
});

export const deletePalette = (paletteId, projectId) => ({
  type: 'DELETE_PALETTE',
  paletteId,
  projectId
});

export const deleteProject = (projectId) => ({
  type: 'DELETE_PROJECT',
  projectId
});

export const expandMiniPalette = (colors) => ({
  type: 'EXPAND_MINI_PALETTE',
  colors
});

export const replacePalette = (palette) => ({
  type: 'REPLACE_PALETTE',
  palette
});

export const updateProjectName = (name) => ({
  type: 'UPDATE_PROJECT_NAME',
  name
});