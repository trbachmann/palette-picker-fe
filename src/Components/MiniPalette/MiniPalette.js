import React from 'react';

const MiniPalette = (props) => {
  const {name, color1, color2, color3, color4, color5 } = props
  return(
    <div>
      <p>{name}</p>
      <div className='MiniPalette--div'>
        <div className='MiniPalette--div--color' style={{backgroundColor: color1}}></div>
        <div className='MiniPalette--div--color' style={{backgroundColor: color2}}></div>
        <div className='MiniPalette--div--color' style={{backgroundColor: color3}}></div>
        <div className='MiniPalette--div--color' style={{backgroundColor: color4}}></div>
        <div className='MiniPalette--div--color' style={{backgroundColor: color5}}></div>
      </div>
    </div>
  )
}

export default MiniPalette;