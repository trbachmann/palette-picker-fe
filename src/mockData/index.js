export const mockProjects = [
  {
    "id": 1,
    "name": "Game Time",
    "created_at": "2019-03-23T18:13:25.614Z",
    "updated_at": "2019-03-23T18:13:25.614Z"
  },
  {
    "id": 2,
    "name": "Movie Tracker",
    "created_at": "2019-03-23T18:13:25.631Z",
    "updated_at": "2019-03-23T18:13:25.631Z"
  },
  {
    "id": 3,
    "name": "Trapper Keeper",
    "created_at": "2019-03-23T18:13:25.652Z",
    "updated_at": "2019-03-23T18:13:25.652Z"
  }
];

export const mockProjectsWithEmptyPalettes = [
  {
    "id": 1,
    "name": "Game Time",
    "created_at": "2019-03-23T18:13:25.614Z",
    "updated_at": "2019-03-23T18:13:25.614Z",
    palettes: []
  },
  {
    "id": 2,
    "name": "Movie Tracker",
    "created_at": "2019-03-23T18:13:25.631Z",
    "updated_at": "2019-03-23T18:13:25.631Z",
    palettes: []
  },
  {
    "id": 3,
    "name": "Trapper Keeper",
    "created_at": "2019-03-23T18:13:25.652Z",
    "updated_at": "2019-03-23T18:13:25.652Z",
    palettes: []
  }
];

export const mockProjectsWithFullPalettes = [
  {
    "id": 1,
    "name": "Game Time",
    "created_at": "2019-03-23T18:13:25.614Z",
    "updated_at": "2019-03-23T18:13:25.614Z",
    palettes: [
      {
          id: 1,
          name: "Not favorites",
          color1: "#FCB97D",
          color2: "#EDD892",
          color3: "#C6B89E",
          color4: "#B5B8A3",
          color5: "#AABA9E",
          project_id: 1,
          created_at: "2019-03-19T20:28:06.194Z",
          updated_at: "2019-03-19T20:28:06.194Z"
      },
      {
          id: 2,
          name: "Favorites",
          color1: "#A0DDFF",
          color2: "#758ECD",
          color3: "#C1CEFE",
          color4: "#7189FF",
          color5: "#624CAB",
          project_id: 1,
          created_at: "2019-03-19T20:28:06.193Z",
          updated_at: "2019-03-19T20:28:06.193Z"
      }
    ]
  },
  {
    "id": 2,
    "name": "Movie Tracker",
    "created_at": "2019-03-23T18:13:25.631Z",
    "updated_at": "2019-03-23T18:13:25.631Z",
    palettes: []
  },
  {
    "id": 3,
    "name": "Trapper Keeper",
    "created_at": "2019-03-23T18:13:25.652Z",
    "updated_at": "2019-03-23T18:13:25.652Z",
    palettes: []
  }
];

export const mockPalette = [
  { hex: "#FCB97D", isLocked: false },
  { hex: "#EDD892", isLocked: false },
  { hex: "#C6B89E", isLocked: false },
  { hex: "#B5B8A3", isLocked: false },
  { hex: "#AABA9E", isLocked: false }
];

export const mockProjectPalettes = [
  {
      id: 1,
      name: "Not favorites",
      color1: "#FCB97D",
      color2: "#EDD892",
      color3: "#C6B89E",
      color4: "#B5B8A3",
      color5: "#AABA9E",
      project_id: 1,
      created_at: "2019-03-19T20:28:06.194Z",
      updated_at: "2019-03-19T20:28:06.194Z"
  },
  {
      id: 2,
      name: "Favorites",
      color1: "#A0DDFF",
      color2: "#758ECD",
      color3: "#C1CEFE",
      color4: "#7189FF",
      color5: "#624CAB",
      project_id: 1,
      created_at: "2019-03-19T20:28:06.193Z",
      updated_at: "2019-03-19T20:28:06.193Z"
  }
];

export const mockProjectsAfterDelete = [
  {
    "id": 1,
    "name": "Game Time",
    "created_at": "2019-03-23T18:13:25.614Z",
    "updated_at": "2019-03-23T18:13:25.614Z",
    palettes: [
      {
        id: 2,
        name: "Favorites",
        color1: "#A0DDFF",
        color2: "#758ECD",
        color3: "#C1CEFE",
        color4: "#7189FF",
        color5: "#624CAB",
        project_id: 1,
        created_at: "2019-03-19T20:28:06.193Z",
        updated_at: "2019-03-19T20:28:06.193Z"
      }
    ]
  },
  {
    "id": 2,
    "name": "Movie Tracker",
    "created_at": "2019-03-23T18:13:25.631Z",
    "updated_at": "2019-03-23T18:13:25.631Z",
    palettes: []
  },
  {
    "id": 3,
    "name": "Trapper Keeper",
    "created_at": "2019-03-23T18:13:25.652Z",
    "updated_at": "2019-03-23T18:13:25.652Z",
    palettes: []
  }
];

export const mockProjectsAfterProjectDelete = [
  {
    "id": 1,
    "name": "Game Time",
    "created_at": "2019-03-23T18:13:25.614Z",
    "updated_at": "2019-03-23T18:13:25.614Z"
  },
  {
    "id": 3,
    "name": "Trapper Keeper",
    "created_at": "2019-03-23T18:13:25.652Z",
    "updated_at": "2019-03-23T18:13:25.652Z"
  }
];

export const mockPaletteBeforeUpdate = {
  name: 'Great Greens',
  color1: "#FCB97D",
  color2: "#EDD892",
  color3: "#C6B89E",
  color4: "#B5B8A3",
  color5: "#AABA9E",
  project_id: 1
};

export const mockPaletteWithUpdates = {
  name: 'Great Green Shades',
  color1: "#GGGGGG",
  color2: "#EDD892",
  color3: "#C6B89E",
  color4: "#G2G2G3",
  color5: "#AABA9E",
  project_id: 1
};

export const mockProjectsBeforePaletteUpdate = [
  {
    id: 1,
    name: "Game Maker",
    created_at: "2019-03-23T18:13:25.614Z",
    updated_at: "2019-03-23T18:13:25.614Z",
    palettes: [
      mockPaletteBeforeUpdate
    ]
  },
  {
    id: 2,
    name: "Keeper Time",
    created_at: "2019-03-23T18:13:27.614Z",
    updated_at: "2019-03-23T18:13:27.614Z",
    palettes: []
  }
];

export const mockProjectsAfterPaletteUpdate = [
  {
    id: 1,
    name: "Game Maker",
    created_at: "2019-03-23T18:13:25.614Z",
    updated_at: "2019-03-23T18:13:25.614Z",
    palettes: [
      mockPaletteWithUpdates
    ]
  },
  {
    id: 2,
    name: "Keeper Time",
    created_at: "2019-03-23T18:13:27.614Z",
    updated_at: "2019-03-23T18:13:27.614Z",
    palettes: []
  }
];

export const mockProjectsAfterNameUpdate = [
  {
    "id": 1,
    "name": "Game Time",
    "created_at": "2019-03-23T18:13:25.614Z",
    "updated_at": "2019-03-23T18:13:25.614Z",
    palettes: []
  },
  {
    "id": 2,
    "name": "Movie Matcher",
    "created_at": "2019-03-23T18:13:25.631Z",
    "updated_at": "2019-03-23T18:13:25.631Z",
    palettes: []
  },
  {
    "id": 3,
    "name": "Trapper Keeper",
    "created_at": "2019-03-23T18:13:25.652Z",
    "updated_at": "2019-03-23T18:13:25.652Z",
    palettes: []
  }
];

export const mockMinPaletteToExpand = [
  "#FGFHES",
  "#E6D292",
  "#B6B89E",
  "#A5B8A3",
  "#DABA9E",
];

export const mockColorsAfterExpand = [
  { hex: "#FGFHES", isLocked: false },
  { hex: "#E6D292", isLocked: false },
  { hex: "#B6B89E", isLocked: false },
  { hex: "#A5B8A3", isLocked: false },
  { hex: "#DABA9E", isLocked: false }
];

export const mockPaletteAfterLock = [
  { hex: "#FCB97D", isLocked: false },
  { hex: "#EDD892", isLocked: false },
  { hex: "#C6B89E", isLocked: false },
  { hex: "#B5B8A3", isLocked: true },
  { hex: "#AABA9E", isLocked: false }
]