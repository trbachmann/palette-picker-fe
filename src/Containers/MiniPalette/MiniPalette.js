import React, { Component } from 'react';
import { connect } from 'react-redux';
import { expandMiniPalette, updatePaletteInEdit } from '../../Actions';

export class MiniPalette extends Component{
  expandPalette = () => {
    const { color1, color2, color3, color4, color5, name, project_id, id } = this.props
    const colors = [color1, color2, color3, color4, color5]
    this.props.expandMiniPalette(colors);
    this.props.updatePaletteInEdit({
      name,
      id,
      color1,
      color2,
      color3,
      color4,
      color5,
      project_id,
    });
  }

  render() {
    const {name, color1, color2, color3, color4, color5 } = this.props
    return(
      <div>
        <div className='MiniPalette-div-controls'>
          <p>{name}</p>
          <i className="fas fa-pencil-alt"></i>
          <i className="far fa-trash-alt"></i>
        </div>
        <div className='MiniPalette--div' onClick={this.expandPalette}>
          <div className='MiniPalette--div--color' style={{backgroundColor: color1}}></div>
          <div className='MiniPalette--div--color' style={{backgroundColor: color2}}></div>
          <div className='MiniPalette--div--color' style={{backgroundColor: color3}}></div>
          <div className='MiniPalette--div--color' style={{backgroundColor: color4}}></div>
          <div className='MiniPalette--div--color' style={{backgroundColor: color5}}></div>
        </div>
      </div>
    )
  };
};

export const mapDispatchToProps = dispatch => ({
  expandMiniPalette: (colors) => dispatch(expandMiniPalette(colors)),
  updatePaletteInEdit: (palette) => dispatch(updatePaletteInEdit(palette))
});

export default connect(null, mapDispatchToProps)(MiniPalette);