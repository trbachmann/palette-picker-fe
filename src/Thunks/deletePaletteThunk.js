import { toggleLoading, setError, deletePalette } from '../Actions';
import { fetchData, createOptions } from '../Utils/fetch';

export const deletePaletteThunk = (paletteId) => {
  return async (dispatch) => {
    const url = `/palettes/${paletteId}`;
    const options = createOptions('DELETE');
    try {
      dispatch(toggleLoading(true));
      await fetchData(url, options);
      dispatch(toggleLoading(false));
      dispatch(deletePalette(paletteId));
    } catch(error) {
      dispatch(toggleLoading(false));
      dispatch(setError(error.message));
    }
  }
}