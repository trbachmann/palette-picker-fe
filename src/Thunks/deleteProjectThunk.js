import { toggleLoading, setError, deleteProject } from '../Actions';
import { fetchData, createOptions } from '../Utils/fetch';

export const deleteProjectThunk = (id) => {
  return async (dispatch) => {
    const url = `/projects/${id}`;
    const options = createOptions('DELETE');
    try {
      dispatch(toggleLoading(true));
      await fetchData(url, options);
      dispatch(toggleLoading(false));
      dispatch(deleteProject(id));
    } catch(error) {
      dispatch(toggleLoading(false));
      dispatch(setError(error.message));
    }
  }
}