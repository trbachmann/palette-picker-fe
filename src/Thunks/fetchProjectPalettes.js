import { addProjectPalettes, setError, toggleLoading } from '../Actions';
import { fetchData } from '../Utils/fetch';

export const fetchProjectPalettes = (id) => {
  return async (dispatch) => {
    try{
      dispatch(toggleLoading(true));
      const projectPalettes = await fetchData(`/projects/${id}/palettes`);
      dispatch(toggleLoading(false));
      dispatch(addProjectPalettes(projectPalettes, id))
    } catch (error) {
      dispatch(toggleLoading(false));
      dispatch(setError(error.message));
    }
  }
};