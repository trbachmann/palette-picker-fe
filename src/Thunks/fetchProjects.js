import { toggleLoading, setError, setProjects } from '../Actions';
import { fetchData } from '../Utils/fetch';

export const fetchProjects = () => {
  return async (dispatch) => {
    try {
      dispatch(toggleLoading(true));
      const projects = await fetchData('/projects');
      dispatch(toggleLoading(false));
      dispatch(setProjects(projects));
    } catch (error) {
      dispatch(setError(error));
    }
  }
};