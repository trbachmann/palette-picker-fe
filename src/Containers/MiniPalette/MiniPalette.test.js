import React from 'react';
import { shallow } from 'enzyme';
import { MiniPalette, mapDispatchToProps } from './MiniPalette';
import * as mockData from '../../mockData';
import { deletePaletteThunk } from '../../Thunks/deletePaletteThunk';
import * as actions from '../../Actions';
import thunk from 'redux-thunk';

jest.mock('../../Thunks/deletePaletteThunk.js');

describe('MiniPalette', () => {
  const mockPalette = mockData.mockProjectPalettes[0];
  const mockColorsOnly = [
    mockPalette.color1,
    mockPalette.color2,
    mockPalette.color3,
    mockPalette.color4,
    mockPalette.color5
  ];
  describe('MiniPalette component', () => {
    let wrapper;
    const mockDeletePaletteThunk = jest.fn();
    const mockExpandMiniPalette = jest.fn();

    beforeEach(() => {
      jest.mock('../../Thunks/deletePaletteThunk.js');
      wrapper = shallow(<MiniPalette {...mockPalette} deletePaletteThunk={mockDeletePaletteThunk} expandMiniPalette={mockExpandMiniPalette} />);
    });
    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should call deletePalette when the icon is clicked', () => {
      const instance = wrapper.instance();
      jest.spyOn(instance, 'deletePalette');
      wrapper.instance().forceUpdate()
      wrapper.find('i').simulate('click');
      expect(instance.deletePalette).toHaveBeenCalled();
    });

    it('should call deletePaletteThunk with id and project id when icon is click', () => {
      wrapper.find('i').simulate('click');
      expect(mockDeletePaletteThunk).toHaveBeenCalledWith(mockPalette.id, mockPalette.project_id);
    });

    it('should call expandPalette when MiniPalette--div--colors is clicked', () => {
      const instance = wrapper.instance();
      jest.spyOn(instance, 'expandPalette');
      wrapper.instance().forceUpdate()
      wrapper.find('.MiniPalette--div--colors').simulate('click');
      expect(instance.expandPalette).toHaveBeenCalled();
    });

    it('should call expandMiniPalette with the colors from the palette when MiniPalette--div--colors is clicked', () => {
      wrapper.find('.MiniPalette--div--colors').simulate('click');
      expect(mockExpandMiniPalette).toHaveBeenCalledWith(mockColorsOnly);
    });
  });

  describe('mapDispatchToProps', () => {
    const mockDispatch = jest.fn();
    it('should call dispatch with expandMiniPalette action when expandMiniPalette is called', () => {
      const actionToDispatch = actions.expandMiniPalette(mockColorsOnly);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.expandMiniPalette(mockColorsOnly);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch with deletPaletteThunk when deletePaletteThunk is called', () => {
      const thunkToDispatch = deletePaletteThunk(mockPalette.id, mockPalette.project_id);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.deletePaletteThunk(mockPalette.id, mockPalette.project_id);
      expect(mockDispatch).toHaveBeenCalledWith(thunkToDispatch);      
    });
  });
});