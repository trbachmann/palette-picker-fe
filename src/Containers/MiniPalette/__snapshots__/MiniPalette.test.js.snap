// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MiniPalette MiniPalette component should match the snapshot 1`] = `
<div>
  <div
    className="MiniPalette--div--controls"
  >
    <p>
      Not favorites
    </p>
    <i
      className="far fa-trash-alt"
      onClick={[Function]}
    />
  </div>
  <div
    className="MiniPalette--div--colors"
    onClick={[Function]}
  >
    <div
      className="MiniPalette--div--color"
      style={
        Object {
          "backgroundColor": "#FCB97D",
        }
      }
    />
    <div
      className="MiniPalette--div--color"
      style={
        Object {
          "backgroundColor": "#EDD892",
        }
      }
    />
    <div
      className="MiniPalette--div--color"
      style={
        Object {
          "backgroundColor": "#C6B89E",
        }
      }
    />
    <div
      className="MiniPalette--div--color"
      style={
        Object {
          "backgroundColor": "#B5B8A3",
        }
      }
    />
    <div
      className="MiniPalette--div--color"
      style={
        Object {
          "backgroundColor": "#AABA9E",
        }
      }
    />
  </div>
</div>
`;
